Replay in nullchain fails on trying to snapshot

Running a replay in vega market sim. 
1. Set up vega market sim following README.md
2. Set vega version in .env to be `07e3753fe001243fa0b0c0e08dd240cb64032330` and run `make` to recompile the binary. 
3. Run `python ./vega_sim/replay/replay.py --retain_log_files  --dir` with the attached replay file here: 
[replay.zip](https://github.com/vegaprotocol/vega/files/9991130/replay.zip) (of course you'll have to unzip first). 
4. You should see the replay ending much too early with 
```
panic: failed to save snapshot to disk

goroutine 22844 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0xc0?, 0x1?, {0x1?, 0x0?, 0x0?})
        /Users/davidsiska/go/pkg/mod/go.uber.org/zap@v1.22.0/zapcore/entry.go:198 +0x7c
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0x1400fc29450, {0x140115619c0, 0x1, 0x1})
        /Users/davidsiska/go/pkg/mod/go.uber.org/zap@v1.22.0/zapcore/entry.go:264 +0x2ac
go.uber.org/zap.(*Logger).Panic(0x1400103c000?, {0x107002736?, 0x140021dd880?}, {0x140115619c0, 0x1, 0x1})
        /Users/davidsiska/go/pkg/mod/go.uber.org/zap@v1.22.0/logger.go:251 +0x54
code.vegaprotocol.io/vega/core/snapshot.(*Engine).snapshotNow.func2()
        /Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/snapshot/engine.go:841 +0x10c
created by code.vegaprotocol.io/vega/core/snapshot.(*Engine).snapshotNow
        /Users/davidsiska/git-repos/github.com/vegaprotocol/vega-market-sim/extern/vega/core/snapshot/engine.go:851 +0xb08
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@
 ### 🛠 Improvements
 - [6795](https://github.com/vegaprotocol/vega/issues/6795 - max gas implementation
 - [6641](https://github.com/vegaprotocol/vega/issues/6641) - network wide limits
+- [6792](https://github.com/vegaprotocol/vega/issues/6792) - Better handling of panics when moving time with `nullchain`, add endpoint to query whether `nullchain` is replaying
 
 ### 🐛 Fixes
 - [6801](https://github.com/vegaprotocol/vega/issues/6801) - Fix internal data source validations

--- a/core/blockchain/nullchain/nullchain.go
+++ b/core/blockchain/nullchain/nullchain.go
@@ -82,6 +82,7 @@ type NullBlockchain struct {
 
 	mu        sync.Mutex
 	replaying atomic.Bool
+	replayer  *Replayer
 }
 
 func NewClient(
@@ -154,6 +155,7 @@ func (n *NullBlockchain) StartChain() error {
 	if err != nil {
 		return err
 	}
+	n.replayer = r
 
 	if n.cfg.Replay.Replay {
 		n.log.Info("nullchain is replaying chain", logging.String("replay-file", n.cfg.Replay.ReplayFile))
@@ -175,7 +177,6 @@ func (n *NullBlockchain) StartChain() error {
 
 	if n.cfg.Replay.Record {
 		n.log.Info("nullchain is recording chain data", logging.String("replay-file", n.cfg.Replay.ReplayFile))
-		n.app = r // sandwich our replayer between nullchain <-> protocol
 	}
 
 	return nil
@@ -186,6 +187,12 @@ func (n *NullBlockchain) processBlock() {
 		n.log.Debugf("processing block %d with %d transactions", n.blockHeight, len(n.pending))
 	}
 
+	resp := abci.ResponseCommit{}
+	if n.replayer != nil && n.cfg.Replay.Record {
+		n.replayer.startBlock(n.blockHeight, n.now.UnixNano(), n.pending)
+		defer func() { n.replayer.saveBlock(resp.Data) }()
+	}
+
 	n.BeginBlock()
 	for _, tx := range n.pending {
 		n.app.DeliverTx(*tx)
@@ -193,7 +200,7 @@ func (n *NullBlockchain) processBlock() {
 	n.pending = n.pending[:0]
 
 	n.EndBlock()
-	n.app.Commit()
+	resp = n.app.Commit()
 
 	// Increment time, blockheight, ready to start a new block
 	n.blockHeight++

--- a/core/blockchain/nullchain/replay.go
+++ b/core/blockchain/nullchain/replay.go
@@ -32,6 +32,7 @@ type Replayer struct {
 	app     ApplicationService
 	rFile   *os.File
 	current *blockData
+	stop    chan struct{}
 }
 
 func NewNullChainReplayer(app ApplicationService, cfg blockchain.ReplayConfig, log *logging.Logger) (*Replayer, error) {
@@ -53,6 +54,7 @@ func NewNullChainReplayer(app ApplicationService, cfg blockchain.ReplayConfig, l
 		app:   app,
 		rFile: f,
 		log:   log,
+		stop:  make(chan struct{}, 1),
 	}, nil
 }
 
@@ -60,43 +62,32 @@ func (r *Replayer) InitChain(req abci.RequestInitChain) (resp abci.ResponseInitC
 	return r.app.InitChain(req)
 }
 
-func (r *Replayer) BeginBlock(req abci.RequestBeginBlock) (resp abci.ResponseBeginBlock) {
-	r.current = &blockData{
-		Height: req.Header.Height,
-		Time:   req.Header.Time.UnixNano(),
-		Txs:    [][]byte{},
-	}
-	return r.app.BeginBlock(req)
-}
-
-func (r *Replayer) EndBlock(req abci.RequestEndBlock) (resp abci.ResponseEndBlock) {
-	return r.app.EndBlock(req)
-}
-
-func (r *Replayer) Commit() (resp abci.ResponseCommit) {
-	resp = r.app.Commit()
-	r.current.AppHash = resp.Data
-	r.write(r.current)
-	r.current = nil
-	return
-}
-
-func (r *Replayer) DeliverTx(req abci.RequestDeliverTx) (resp abci.ResponseDeliverTx) {
-	r.current.Txs = append(r.current.Txs, req.Tx)
-	return r.app.DeliverTx(req)
-}
-
-func (r *Replayer) Info(req abci.RequestInfo) (resp abci.ResponseInfo) {
-	return r.app.Info(req)
-}
-
 func (r *Replayer) Stop() error {
-	if r.current != nil {
-		// a panic must've occurred while processing a block because we didn't make it to commit
-		// save what we have
-		r.write(r.current)
-	}
+	r.stop <- struct{}{}
+	close(r.stop)
 	return r.rFile.Close()
+}
+
+// startBlock saves in memory all the transactions in the block, we do not write until saveBlock us called
+// with a potential appHash.
+func (r *Replayer) startBlock(height, now int64, txs []*abci.RequestDeliverTx) {
+	r.current = &blockData{
+		Height: height,
+		Time:   now,
+	}
+	for _, tx := range txs {
+		r.current.Txs = append(r.current.Txs, tx.Tx)
+	}
+}
+
+// saveBlock writes to the replay file the details of the current block adding the appHash to it.
+// If a panic occurred appHash may be empty.
+func (r *Replayer) saveBlock(appHash []byte) {
+	r.current.AppHash = appHash
+	if err := r.write(); err != nil {
+		r.log.Panic("unable to write block to file", logging.Int64("block-height", r.current.Height))
+	}
+	r.current = nil
 }
 
 // replayChain sends all the recorded per-block transactions into the protocol returning the block-height and block-time it reached
@@ -109,6 +100,13 @@ func (r *Replayer) replayChain(appHeight int64, chainID string) (int64, time.Tim
 	var replayedHeight int64
 	var replayedTime time.Time
 	for s.Scan() {
+		select {
+		case <-r.stop:
+			r.log.Info("core is shutting down, nullchain replaying stopped", logging.Int64("block-height", replayedHeight))
+			return replayedHeight, replayedTime, nil
+		default:
+		}
+
 		var data blockData
 		if err := json.Unmarshal(s.Bytes(), &data); err != nil {
 			return replayedHeight, replayedTime, err
@@ -165,10 +163,10 @@ func (r *Replayer) replayChain(appHeight int64, chainID string) (int64, time.Tim
 	return replayedHeight, replayedTime, nil
 }
 
-func (r *Replayer) write(bd *blockData) error {
+func (r *Replayer) write() error {
 	b, err := json.Marshal(r.current)
 	if err != nil {
-		return fmt.Errorf("unable to record block %d: %w", bd.Height, err)
+		return fmt.Errorf("unable to record block %d: %w", r.current.Height, err)
 	}
 
 	// write each marshalled json block on a new line, its crude, but lets worry about perf if perf becomes a problem.

--- a/core/blockchain/nullchain/server.go
+++ b/core/blockchain/nullchain/server.go
@@ -28,9 +28,14 @@ import (
 
 var ErrInvalidRequest = errors.New("invalid request")
 
+const (
+	NullChainStatusReady     = "chain-ready"
+	NullChainStatusReplaying = "chain-replaying"
+)
+
 func (n *NullBlockchain) Stop() error {
-	if r, ok := n.app.(*Replayer); ok {
-		if err := r.Stop(); err != nil {
+	if n.replayer != nil {
+		if err := n.replayer.Stop(); err != nil {
 			n.log.Error("failed to stop nullchain replayer")
 		}
 	}
@@ -53,16 +58,18 @@ func (n *NullBlockchain) Start() error {
 }
 
 func (n *NullBlockchain) StartServer() error {
+	n.srv = &http.Server{Addr: net.JoinHostPort(n.cfg.IP, strconv.Itoa(n.cfg.Port))}
+	http.HandleFunc("/api/v1/forwardtime", n.handleForwardTime)
+	http.HandleFunc("/api/v1/status", n.status)
+
+	n.log.Info("starting time-forwarding server", logging.String("addr", n.srv.Addr))
+	go n.srv.ListenAndServe()
+
 	n.log.Info("starting blockchain")
 	if err := n.StartChain(); err != nil {
 		return err
 	}
 
-	n.srv = &http.Server{Addr: net.JoinHostPort(n.cfg.IP, strconv.Itoa(n.cfg.Port))}
-	http.HandleFunc("/api/v1/forwardtime", n.handleForwardTime)
-
-	n.log.Info("starting time-forwarding server", logging.String("addr", n.srv.Addr))
-	go n.srv.ListenAndServe()
 	return nil
 }
 
@@ -110,6 +117,45 @@ func (n *NullBlockchain) handleForwardTime(w http.ResponseWriter, r *http.Reques
 		http.Error(w, err.Error(), http.StatusBadRequest)
 	}
 
-	// Do the dance
-	n.ForwardTime(d)
+	if n.replaying.Load() {
+		http.Error(w, ErrChainReplaying.Error(), http.StatusServiceUnavailable)
+	}
+
+	// we need to call ForwardTime in a different routine so that if it panics it stops the node instead of just being
+	// caught in the http-handler recover. But awkwardly it seems like the vega-sim relies on the http-request not
+	// returning until the time-forward has finished, so we need to preserve that for now.
+	done := make(chan struct{})
+	go func() {
+		n.ForwardTime(d)
+		done <- struct{}{}
+	}()
+	<-done
+}
+
+// status returns the status of the nullchain, whether it is replaying or whether its ready to go.
+func (n *NullBlockchain) status(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+
+	resp := struct {
+		Status string `json:"status"`
+	}{
+		Status: NullChainStatusReady,
+	}
+
+	if n.replaying.Load() {
+		resp.Status = NullChainStatusReplaying
+	}
+
+	buf, err := json.Marshal(resp)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	_, err = w.Write(buf)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
 }

